lib/stream/promises.js

```diff
@@ -23,11 +23,15 @@ function pipeline(...streams) {
       signal = options.signal;
     }
 
-    pl(streams, (err, value) => {
+    const stream = pl(streams, (err, value) => {
       if (err) {
         reject(err);
-      } else {
+      } else if (value !== undefined) {
         resolve(value);
+      } else if (stream.readable) {
+        resolve(stream);
```

Comment:
I'm not sure I understand. Does this cover the case where the last stream in the pipeline is a `Transform`? If so I think it does not make much sense as the `pipeline()` callback might never be called and the promise resolved if the stream is not read. If not, can it be readable?

Reply:
If the last stream is readable then pipeline only waits until it is no longer writable, it doens't wait for readable to complete.

Reply:
Ok but if the writable side depends on the readable side and the readable is paused and is not resumed it will wait indefinitely, no? It is still writable but it does not progress.

Reply:
> writable side depends on the readable side

When does that happen? Also I'm not sure how this is relevant for this PR? That's already a problem with pipeline in that case.

Reply:
> When does that happen?

```js
const stream = require('stream');

const chunk = Buffer.alloc(1024);
let bytesRead = 0;

const readable = new stream.Readable({
  read() {
    if (bytesRead === 100 * 1024) {
      readable.push(null);
    } else {
      bytesRead += chunk.length;
      readable.push(chunk);
    }
  }
});

const passThrough = new stream.PassThrough();

stream.pipeline(readable, passThrough, function (err) {
  if (err) {
    throw err;
  }

  console.log('done');
});
```

> Also I'm not sure how this is relevant for this PR? That's already a problem with pipeline in that case.

Yes, I agree, it's already a problem with pipeline because I think it does not make sense to have a readable destination due to how pipeline works.

It is relevant to this PR because the promise is not resolved. No stream is returned contrary to what commit message says.

Reply:
If this PR is only for the case where the pipeline promise is resolved, then it should be good but I think that in that case the destination stream is never paused or the user has direct control over it, so they can already read any remaining data.

Reply:
> If this PR is only for the case where the pipeline promise is resolved, then it should be good but I think that in that case the destination stream is never paused or the user has direct control over it, so they can already read any remaining data.

Not true if the last entry is a generator.

Reply:
Ok, in that case the user does not control the internally created `PassThrough` proxy but I think it does not change much:

```js
const stream = require('stream');

const chunk = Buffer.alloc(1024);
let bytesRead = 0;

const readable = new stream.Readable({
  read() {
    if (bytesRead === 100 * 1024) {
      readable.push(null);
    } else {
      bytesRead += chunk.length;
      readable.push(chunk);
    }
  }
});

async function* passThrough(source) {
  for await (const chunk of source) {
    yield chunk;
  }
}

stream.pipeline(readable, passThrough, function (err) {
  if (err) {
    throw err;
  }

  console.log('done');
});
```

The promise is still not resolved so the user cannot read from the proxy unless it is not paused.

If the PR was made to solve this specific use case (read buffered data from passthrough proxy when promise is resolved) then I'm fine with it but I would appreciate a summary of what we discussed here in commit message.

Reply:
I still don't see the relevance to this PR. 😕

Reply:
My whole point is that it does not make sense to use a `Transform` as destination as it could hang the pipeline (especially in the promisified version). In my opinion this is a fix for an edge case that should not exist in the first place :)

Anyway I have no objections.